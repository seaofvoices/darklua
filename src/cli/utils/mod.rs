pub fn maybe_plural(count: usize) -> &'static str {
    if count > 1 {
        "s"
    } else {
        ""
    }
}
