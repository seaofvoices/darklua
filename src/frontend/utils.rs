use std::{
    path::{Component, Path, PathBuf},
    time::{Duration, Instant},
};

pub fn maybe_plural(count: usize) -> &'static str {
    if count > 1 {
        "s"
    } else {
        ""
    }
}

pub fn normalize_path(path: &Path) -> PathBuf {
    let mut components = path.components().peekable();
    let mut ret = if let Some(c @ Component::Prefix(..)) = components.peek().cloned() {
        components.next();
        PathBuf::from(c.as_os_str())
    } else {
        PathBuf::new()
    };

    for component in components {
        match component {
            Component::Prefix(..) => unreachable!(),
            Component::RootDir => {
                ret.push(component.as_os_str());
            }
            Component::CurDir => {}
            Component::ParentDir => {
                ret.pop();
            }
            Component::Normal(c) => {
                ret.push(c);
            }
        }
    }
    ret
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct Timer {
    start: Instant,
    accumulated_time: Duration,
}

impl Timer {
    pub fn now() -> Self {
        Self {
            start: Instant::now(),
            accumulated_time: Duration::new(0, 0),
        }
    }

    pub fn pause(&mut self) {
        let duration = self.start.elapsed();
        self.accumulated_time += duration;
    }

    pub fn start(&mut self) {
        self.start = Instant::now();
    }

    pub fn duration_label(&self) -> String {
        let duration = self.start.elapsed();
        durationfmt::to_string(duration + self.accumulated_time)
    }
}
