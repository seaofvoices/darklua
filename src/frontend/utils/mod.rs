#[cfg(not(target_arch = "wasm32"))]
mod timer;
#[cfg(target_arch = "wasm32")]
mod wasm_timer;

#[cfg(not(target_arch = "wasm32"))]
pub use timer::Timer;
#[cfg(target_arch = "wasm32")]
pub use wasm_timer::Timer;

use std::path::{Component, Path, PathBuf};

pub fn maybe_plural(count: usize) -> &'static str {
    if count > 1 {
        "s"
    } else {
        ""
    }
}

pub fn normalize_path(path: &Path) -> PathBuf {
    let mut components = path.components().peekable();
    let mut ret = if let Some(c @ Component::Prefix(..)) = components.peek().cloned() {
        components.next();
        PathBuf::from(c.as_os_str())
    } else {
        PathBuf::new()
    };

    for component in components {
        match component {
            Component::Prefix(..) => unreachable!(),
            Component::RootDir => {
                ret.push(component.as_os_str());
            }
            Component::CurDir => {}
            Component::ParentDir => {
                ret.pop();
            }
            Component::Normal(c) => {
                ret.push(c);
            }
        }
    }
    ret
}
