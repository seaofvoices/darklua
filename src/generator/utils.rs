//! A module that contains the main [LuaGenerator](trait.LuaGenerator.html) trait
//! and its implementations.

use crate::nodes::{
    Expression, FieldExpression, FunctionCall, IndexExpression, NumberExpression, Prefix,
    Statement, StringExpression, Variable,
};

const QUOTED_STRING_MAX_LENGTH: usize = 60;
const LONG_STRING_MIN_LENGTH: usize = 20;
const FORCE_LONG_STRING_NEW_LINE_THRESHOLD: usize = 6;

pub fn is_relevant_for_spacing(character: &char) -> bool {
    character.is_ascii_alphabetic() || character.is_ascii_digit() || *character == '_'
}

pub fn break_long_string(last_str: &str) -> bool {
    if let Some(last_char) = last_str.chars().last() {
        last_char == '['
    } else {
        false
    }
}

pub fn break_variable_arguments(last_string: &str) -> bool {
    if let Some('.') = last_string.chars().last() {
        true
    } else if let Some(first_char) = last_string.chars().next() {
        first_char == '.' || first_char.is_ascii_digit()
    } else {
        false
    }
}

pub fn break_minus(last_string: &str) -> bool {
    if let Some(last_char) = last_string.chars().last() {
        last_char == '-'
    } else {
        false
    }
}

pub fn break_concat(last_string: &str) -> bool {
    if let Some('.') = last_string.chars().last() {
        true
    } else if let Some(first_char) = last_string.chars().next() {
        first_char == '.' || first_char.is_ascii_digit()
    } else {
        false
    }
}

pub fn ends_with_prefix(statement: &Statement) -> bool {
    match statement {
        Statement::Assign(assign) => {
            if let Some(value) = assign.last_value() {
                expression_ends_with_prefix(value)
            } else {
                false
            }
        }
        Statement::CompoundAssign(assign) => expression_ends_with_prefix(assign.get_value()),
        Statement::Call(_) => true,
        Statement::Repeat(repeat) => expression_ends_with_prefix(repeat.get_condition()),
        Statement::LocalAssign(assign) => {
            if let Some(value) = assign.last_value() {
                expression_ends_with_prefix(value)
            } else {
                false
            }
        }
        _ => false,
    }
}

pub fn starts_with_parenthese(statement: &Statement) -> bool {
    match statement {
        Statement::Assign(assign) => {
            if let Some(variable) = assign.get_variables().first() {
                match variable {
                    Variable::Identifier(_) => false,
                    Variable::Field(field) => field_starts_with_parenthese(field),
                    Variable::Index(index) => index_starts_with_parenthese(index),
                }
            } else {
                false
            }
        }
        Statement::CompoundAssign(assign) => match assign.get_variable() {
            Variable::Identifier(_) => false,
            Variable::Field(field) => field_starts_with_parenthese(field),
            Variable::Index(index) => index_starts_with_parenthese(index),
        },
        Statement::Call(call) => call_starts_with_parenthese(call),
        _ => false,
    }
}

fn expression_ends_with_prefix(expression: &Expression) -> bool {
    match expression {
        Expression::Binary(binary) => expression_ends_with_prefix(binary.right()),
        Expression::Call(_)
        | Expression::Parenthese(_)
        | Expression::Identifier(_)
        | Expression::Field(_)
        | Expression::Index(_) => true,
        Expression::Unary(unary) => expression_ends_with_prefix(unary.get_expression()),
        Expression::If(if_expression) => {
            expression_ends_with_prefix(if_expression.get_else_result())
        }
        Expression::False(_)
        | Expression::Function(_)
        | Expression::Nil(_)
        | Expression::Number(_)
        | Expression::String(_)
        | Expression::Table(_)
        | Expression::True(_)
        | Expression::VariableArguments(_) => false,
    }
}

fn prefix_starts_with_parenthese(prefix: &Prefix) -> bool {
    match prefix {
        Prefix::Parenthese(_) => true,
        Prefix::Call(call) => call_starts_with_parenthese(call),
        Prefix::Field(field) => field_starts_with_parenthese(field),
        Prefix::Index(index) => index_starts_with_parenthese(index),
        Prefix::Identifier(_) => false,
    }
}

#[inline]
fn call_starts_with_parenthese(call: &FunctionCall) -> bool {
    prefix_starts_with_parenthese(call.get_prefix())
}

#[inline]
fn field_starts_with_parenthese(field: &FieldExpression) -> bool {
    prefix_starts_with_parenthese(field.get_prefix())
}

#[inline]
fn index_starts_with_parenthese(index: &IndexExpression) -> bool {
    prefix_starts_with_parenthese(index.get_prefix())
}

pub fn write_number(number: &NumberExpression) -> String {
    match number {
        NumberExpression::Decimal(number) => {
            let float = number.get_raw_float();
            if float.is_nan() {
                "(0/0)".to_owned()
            } else if float.is_infinite() {
                format!("({}1/0)", if float.is_sign_negative() { "-" } else { "" })
            } else {
                format!(
                    "{:.}{}",
                    float,
                    number
                        .get_exponent()
                        .map(|exponent| {
                            let exponent_char = number
                                .is_uppercase()
                                .map(|is_uppercase| if is_uppercase { 'E' } else { 'e' })
                                .unwrap_or('e');
                            format!("{}{}", exponent_char, exponent)
                        })
                        .unwrap_or_else(|| "".to_owned())
                )
            }
        }
        NumberExpression::Hex(number) => {
            format!(
                "0{}{:x}{}",
                if number.is_x_uppercase() { 'X' } else { 'x' },
                number.get_raw_integer(),
                number
                    .get_exponent()
                    .map(|exponent| {
                        let exponent_char = number
                            .is_exponent_uppercase()
                            .map(|is_uppercase| if is_uppercase { 'P' } else { 'p' })
                            .unwrap_or('p');
                        format!("{}{}", exponent_char, exponent)
                    })
                    .unwrap_or_else(|| "".to_owned())
            )
        }
        NumberExpression::Binary(number) => {
            format!(
                "0{}{:b}",
                if number.is_b_uppercase() { 'B' } else { 'b' },
                number.get_raw_value()
            )
        }
    }
}

fn needs_escaping(character: char) -> bool {
    !(character.is_ascii_graphic() || character == ' ') || character == '\\'
}

fn needs_quoted_string(character: char) -> bool {
    !(character.is_ascii_graphic() || character == ' ' || character == '\n')
}

fn escape(character: char) -> String {
    match character {
        '\n' => "\\n".to_owned(),
        '\t' => "\\t".to_owned(),
        '\\' => "\\\\".to_owned(),
        '\r' => "\\r".to_owned(),
        '\u{7}' => "\\a".to_owned(),
        '\u{8}' => "\\b".to_owned(),
        '\u{B}' => "\\v".to_owned(),
        '\u{C}' => "\\f".to_owned(),
        _ => {
            if character.is_ascii() {
                format!("\\{}", character as u8)
            } else {
                format!("\\u{{{:x}}}", character as u32)
            }
        }
    }
}

#[inline]
pub fn count_new_lines(string: &str) -> usize {
    string.chars().filter(|c| *c == '\n').count()
}

pub fn write_string(string: &StringExpression) -> String {
    let value = string.get_value();

    if value.is_empty() {
        return "''".to_owned();
    }

    if value.len() == 1 {
        let character = value
            .chars()
            .next()
            .expect("string should have at least one character");
        match character {
            '\'' => return "\"'\"".to_owned(),
            '"' => return "'\"'".to_owned(),
            _ => {
                if needs_escaping(character) {
                    return format!("'{}'", escape(character));
                } else {
                    return format!("'{}'", character);
                }
            }
        }
    }

    if !value.contains(needs_quoted_string)
        && value.len() >= LONG_STRING_MIN_LENGTH
        && (value.len() >= QUOTED_STRING_MAX_LENGTH
            || count_new_lines(value) >= FORCE_LONG_STRING_NEW_LINE_THRESHOLD)
    {
        write_long_bracket(value)
    } else {
        write_quoted(value)
    }
}

fn write_long_bracket(value: &str) -> String {
    let mut i = if value.ends_with(']') { 1 } else { 0 };
    let mut equals = "=".repeat(i);
    loop {
        if !value.contains(&format!("]{}]", equals)) {
            break;
        } else {
            i += 1;
            equals = "=".repeat(i);
        };
    }
    let needs_extra_new_line = if value.starts_with('\n') { "\n" } else { "" };
    format!("[{}[{}{}]{}]", equals, needs_extra_new_line, value, equals)
}

fn write_quoted(value: &str) -> String {
    let mut quoted = String::new();
    quoted.reserve(value.len() + 2);

    let quote_symbol = get_quote_symbol(value);
    quoted.push(quote_symbol);

    for character in value.chars() {
        if character == quote_symbol {
            quoted.push('\\');
            quoted.push(quote_symbol);
        } else if needs_escaping(character) {
            quoted.push_str(&escape(character));
        } else {
            quoted.push(character);
        }
    }

    quoted.push(quote_symbol);
    quoted.shrink_to_fit();
    quoted
}

fn get_quote_symbol(value: &str) -> char {
    if value.contains('"') {
        '\''
    } else if value.contains('\'') {
        '"'
    } else {
        '\''
    }
}

#[cfg(test)]
mod test {
    use super::*;

    mod write_string {
        use super::*;

        macro_rules! test_output {
            ($($name:ident($input:literal) => $value:literal),* $(,)?) => {
                $(
                    #[test]
                    fn $name() {
                        assert_eq!($value, write_string(&StringExpression::from_value($input)));
                    }
                )*
            };
        }

        test_output!(
            empty("") => "''",
            single_letter("a") => "'a'",
            single_digit("8") => "'8'",
            single_symbol("!") => "'!'",
            single_space(" ") => "' '",
            abc("abc") => "'abc'",
            three_spaces("   ") => "'   '",
            new_line("\n") => "'\\n'",
            bell("\u{7}") => "'\\a'",
            backspace("\u{8}") => "'\\b'",
            form_feed("\u{c}") => "'\\f'",
            tab("\t") => "'\\t'",
            carriage_return("\u{D}") => "'\\r'",
            vertical_tab("\u{B}") => "'\\v'",
            backslash("\\") => "'\\\\'",
            single_quote("'") => "\"'\"",
            double_quote("\"") => "'\"'",
            null("\0") => "'\\0'",
            escape("\u{1B}") => "'\\27'",
            unicode("\u{10FFFF}") => "'\\u{10ffff}'",
            im_cool("I'm cool") => "\"I'm cool\"",
            ends_with_closing_bracket("oof]") => "'oof]'",
            multiline_ends_with_closing_bracket("oof\noof]") => "'oof\\noof]'",
            large_multiline_does_not_end_with_closing_bracket("ooof\nooof\nooof\nooof\nooof\nooof\nooof\nooof\noof")
                => "[[ooof\nooof\nooof\nooof\nooof\nooof\nooof\nooof\noof]]",
            large_multiline_ends_with_closing_bracket("ooof\nooof\nooof\nooof\nooof\nooof\nooof\nooof\noof]")
                => "[=[ooof\nooof\nooof\nooof\nooof\nooof\nooof\nooof\noof]]=]",
            large_multiline_starts_with_new_line("\nooof\nooof\nooof\nooof\nooof\nooof\nooof\nooof\noof")
                => "[[\n\nooof\nooof\nooof\nooof\nooof\nooof\nooof\nooof\noof]]",

            large_multiline_with_unicode("\nooof\nooof\nooof\nooof\nooof\nooof\nooof\nooof\noof\u{10FFFF}")
                => "'\\nooof\\nooof\\nooof\\nooof\\nooof\\nooof\\nooof\\nooof\\noof\\u{10ffff}'",
        );
    }
}
