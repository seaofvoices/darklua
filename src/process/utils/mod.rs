mod permutator;

pub(crate) use permutator::Permutator;

pub(crate) type CharPermutator = Permutator<std::str::Chars<'static>>;

pub(crate) fn identifier_permutator() -> CharPermutator {
    Permutator::new("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_0123456789".chars())
}

pub(crate) const KEYWORDS: [&str; 21] = [
    "and", "break", "do", "else", "elseif", "end", "false", "for", "function", "if", "in", "local",
    "nil", "not", "or", "repeat", "return", "then", "true", "until", "while",
];

macro_rules! matches_any_keyword {
    () => {
        "and"
            | "break"
            | "do"
            | "else"
            | "elseif"
            | "end"
            | "false"
            | "for"
            | "function"
            | "if"
            | "in"
            | "local"
            | "nil"
            | "not"
            | "or"
            | "repeat"
            | "return"
            | "then"
            | "true"
            | "until"
            | "while"
    };
}

pub(crate) fn is_valid_identifier(identifier: &str) -> bool {
    !identifier.is_empty()
        && identifier.is_ascii()
        && identifier
            .char_indices()
            .all(|(i, c)| c.is_alphabetic() || c == '_' || (c.is_ascii_digit() && i > 0))
        && !matches!(identifier, matches_any_keyword!())
}

#[cfg(test)]
mod test {
    use super::*;
    #[test]
    fn is_valid_identifier_is_true() {
        assert!(is_valid_identifier("hello"));
        assert!(is_valid_identifier("foo"));
        assert!(is_valid_identifier("bar"));
        assert!(is_valid_identifier("VAR"));
        assert!(is_valid_identifier("_VAR"));
        assert!(is_valid_identifier("_0"));
    }

    #[test]
    fn is_valid_identifier_is_false() {
        assert!(!is_valid_identifier(""));
        assert!(!is_valid_identifier("$hello"));
        assert!(!is_valid_identifier(" "));
        assert!(!is_valid_identifier("5"));
        assert!(!is_valid_identifier("1bar"));
        assert!(!is_valid_identifier("var "));
        assert!(!is_valid_identifier("sp ace"));
    }

    #[test]
    fn keywords_are_not_valid_identifiers() {
        for keyword in KEYWORDS {
            assert!(!is_valid_identifier(keyword));
        }
    }
}
