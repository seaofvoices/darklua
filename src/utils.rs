use std::{
    ffi::OsStr,
    iter::FromIterator,
    path::{Component, Path, PathBuf},
};

pub(crate) fn normalize_path(path: impl AsRef<Path>) -> PathBuf {
    let mut components = path.as_ref().components().peekable();
    let mut ret = if let Some(c @ Component::Prefix(..)) = components.peek().cloned() {
        components.next();
        vec![c.as_os_str()]
    } else {
        Vec::new()
    };

    for component in components {
        match component {
            Component::Prefix(..) => unreachable!(),
            Component::RootDir => {
                ret.push(component.as_os_str());
            }
            Component::CurDir => {}
            Component::ParentDir => {
                if ret.last().filter(|c| **c != OsStr::new("..")).is_some() {
                    ret.pop();
                } else {
                    ret.push(OsStr::new(".."))
                }
            }
            Component::Normal(c) => {
                ret.push(c);
            }
        }
    }

    PathBuf::from_iter(ret)
}

#[cfg(test)]
mod test {
    use super::*;

    fn verify_normalize_path(input: impl AsRef<Path>, output: impl AsRef<Path>) {
        assert_eq!(normalize_path(input.as_ref()), output.as_ref());
    }

    #[test]
    fn current_directory_with_name() {
        verify_normalize_path("./directory", "directory")
    }

    #[test]
    fn from_directory() {
        verify_normalize_path("/directory", "/directory")
    }

    #[test]
    fn parent_directory_path() {
        verify_normalize_path("..", "..")
    }

    #[test]
    fn src_parent_with_directory_name() {
        verify_normalize_path("src/../directory", "directory")
    }

    #[test]
    fn double_parent_directory_path() {
        verify_normalize_path("../..", "../..")
    }
}
