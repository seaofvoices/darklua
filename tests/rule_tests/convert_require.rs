use darklua_core::{rules::Rule, Options, Resources};

use crate::utils;

use super::memory_resources;

test_rule!(
    convert_path_require_to_roblox,
    json5::from_str::<Box<dyn Rule>>(
        r#"{
            rule: 'convert_require',
            current: 'path',
            target: { name: 'roblox', indexing_style: 'find_first_child' },
        }"#
    ).unwrap(),
    resources = memory_resources!(
        "src/test/init.lua" => "return nil",
        "src/test/init.luau" => "return nil",
        "src/test/module.lua" => "return nil",
        "src/test/module.luau" => "return nil",
        "src/test/folder/lib.lua" => "return nil",
        "src/sub/lib.lua" => "return nil",
        "src/format.lua" => "return nil",
        "project.lua" => "return nil",
    ),
    test_file_name = "src/test/runner.lua",
    sibling_module("local module = require('./module.lua')")
        => "local module = require(script.Parent:FindFirstChild('module'))",
    sibling_luau_module("local module = require('./module.luau')")
        => "local module = require(script.Parent:FindFirstChild('module'))",
    sibling_folder_module("local module = require('./module')")
        => "local module = require(script.Parent:FindFirstChild('module'))",
    sibling_module_is_init_file_with_extension("local module = require('./init.lua')")
        => "local module = require(script.Parent)",
    sibling_module_is_init_file_with_luau_extension("local module = require('./init.luau')")
        => "local module = require(script.Parent)",
    module_nested_in_sibling_folder("local module = require('./folder/lib.lua')")
        => "local module = require(script.Parent:FindFirstChild('folder'):FindFirstChild('lib'))",
    module_in_parent("local module = require('../format.lua')")
        => "local module = require(script.Parent.Parent:FindFirstChild('format'))",
    module_nested_in_folder_from_parent("local module = require('../sub/lib.lua')")
        => "local module = require(script.Parent.Parent:FindFirstChild('sub'):FindFirstChild('lib'))",
    module_in_double_parent("local module = require('../../project.lua')")
        => "local module = require(script.Parent.Parent.Parent:FindFirstChild('project'))",
    module_in_parent_with_current_dir("local module = require('.././format.lua')")
        => "local module = require(script.Parent.Parent:FindFirstChild('format'))",
);

test_rule!(
    convert_path_require_to_roblox_with_wait_for_child,
    json5::from_str::<Box<dyn Rule>>(
        r#"{
            rule: 'convert_require',
            current: 'path',
            target: { name: 'roblox', indexing_style: 'wait_for_child' },
        }"#
    ).unwrap(),
    resources = memory_resources!(
        "src/test/init.lua" => "return nil",
        "src/test/init.luau" => "return nil",
        "src/test/module.lua" => "return nil",
        "src/test/module.luau" => "return nil",
        "src/test/folder/lib.lua" => "return nil",
        "src/sub/lib.lua" => "return nil",
        "src/format.lua" => "return nil",
        "project.lua" => "return nil",
    ),
    test_file_name = "src/test/runner.lua",
    sibling_module("local module = require('./module.lua')")
        => "local module = require(script.Parent:WaitForChild('module'))",
    sibling_luau_module("local module = require('./module.luau')")
        => "local module = require(script.Parent:WaitForChild('module'))",
    sibling_folder_module("local module = require('./module')")
        => "local module = require(script.Parent:WaitForChild('module'))",
    sibling_module_is_init_file_with_extension("local module = require('./init.lua')")
        => "local module = require(script.Parent)",
    sibling_module_is_init_file_with_luau_extension("local module = require('./init.luau')")
        => "local module = require(script.Parent)",
    module_nested_in_sibling_folder("local module = require('./folder/lib.lua')")
        => "local module = require(script.Parent:WaitForChild('folder'):WaitForChild('lib'))",
    module_in_parent("local module = require('../format.lua')")
        => "local module = require(script.Parent.Parent:WaitForChild('format'))",
    module_nested_in_folder_from_parent("local module = require('../sub/lib.lua')")
        => "local module = require(script.Parent.Parent:WaitForChild('sub'):WaitForChild('lib'))",
    module_in_double_parent("local module = require('../../project.lua')")
        => "local module = require(script.Parent.Parent.Parent:WaitForChild('project'))",
    module_in_parent_with_current_dir("local module = require('.././format.lua')")
        => "local module = require(script.Parent.Parent:WaitForChild('format'))",
);

test_rule!(
    convert_path_require_to_roblox_with_property_index,
    json5::from_str::<Box<dyn Rule>>(
        r#"{
            rule: 'convert_require',
            current: 'path',
            target: { name: 'roblox', indexing_style: 'property' },
        }"#
    ).unwrap(),
    resources = memory_resources!(
        "src/test/init.lua" => "return nil",
        "src/test/init.luau" => "return nil",
        "src/test/module.lua" => "return nil",
        "src/test/module.luau" => "return nil",
        "src/test/folder/lib.lua" => "return nil",
        "src/sub/lib.lua" => "return nil",
        "src/format.lua" => "return nil",
        "project.lua" => "return nil",
        // specific to property index style tests
        "src/test/while.lua" => "return nil",
        "src/test/a module.lua" => "return nil",
    ),
    test_file_name = "src/test/runner.lua",
    sibling_module("local module = require('./module.lua')")
        => "local module = require(script.Parent.module)",
    sibling_luau_module("local module = require('./module.luau')")
        => "local module = require(script.Parent.module)",
    sibling_folder_module("local module = require('./module')")
        => "local module = require(script.Parent.module)",
    sibling_module_is_init_file_with_extension("local module = require('./init.lua')")
        => "local module = require(script.Parent)",
    sibling_module_is_init_file_with_luau_extension("local module = require('./init.luau')")
        => "local module = require(script.Parent)",
    module_nested_in_sibling_folder("local module = require('./folder/lib.lua')")
        => "local module = require(script.Parent.folder.lib)",
    module_in_parent("local module = require('../format.lua')")
        => "local module = require(script.Parent.Parent.format)",
    module_nested_in_folder_from_parent("local module = require('../sub/lib.lua')")
        => "local module = require(script.Parent.Parent.sub.lib)",
    module_in_double_parent("local module = require('../../project.lua')")
        => "local module = require(script.Parent.Parent.Parent.project)",
    module_in_parent_with_current_dir("local module = require('.././format.lua')")
        => "local module = require(script.Parent.Parent.format)",
    // specific to property index style
    sibling_module_with_keyword_name("local module = require('./while.lua')")
        => "local module = require(script.Parent['while'])",
    sibling_module_with_name_with_space("local module = require('./a module.lua')")
        => "local module = require(script.Parent['a module'])",
);

fn process_file(resources: &Resources, file_name: &str) -> String {
    darklua_core::process(resources, Options::new(file_name))
        .result()
        .unwrap();

    resources.get(file_name).unwrap()
}

fn expect_file_process(resources: &Resources, file_name: &str, expect_content: &str) {
    pretty_assertions::assert_eq!(process_file(resources, file_name), expect_content);
}

fn snapshot_file_process(resources: &Resources, file_name: &str, snapshot_name: &str) {
    insta::assert_snapshot!(
        snapshot_name,
        process_file(resources, file_name),
        &format!("process `tests/test_cases/sourcemap/{}`", file_name)
    );
}

const CONVERT_PATH_TO_ROBLOX_DEFAULT_CONFIG: &str =
    "{ rules: [{ rule: 'convert_require', current: 'path', target: 'roblox' }], generator: \"retain_lines\" }";

#[test]
fn convert_sibling_module_from_init_module() {
    let resources = memory_resources!(
        "src/init.lua" => "local value = require('./value.lua')",
        "src/value.lua" => "return nil",
        ".darklua.json" => CONVERT_PATH_TO_ROBLOX_DEFAULT_CONFIG,
    );
    expect_file_process(
        &resources,
        "src/init.lua",
        "local value = require(script:FindFirstChild('value'))",
    );
}

#[test]
fn convert_sibling_init_module_from_init_module() {
    let resources = memory_resources!(
        "src/init.lua" => "local value = require('./folder/init.lua')",
        "src/folder/init.lua" => "return nil",
        ".darklua.json" => CONVERT_PATH_TO_ROBLOX_DEFAULT_CONFIG,
    );
    expect_file_process(
        &resources,
        "src/init.lua",
        "local value = require(script:FindFirstChild('folder'))",
    );
}

#[test]
fn convert_parent_init_module_from_init_module() {
    let resources = memory_resources!(
        "src/module/init.lua" => "local value = require('../init.lua')",
        "src/init.lua" => "return nil",
        ".darklua.json" => CONVERT_PATH_TO_ROBLOX_DEFAULT_CONFIG,
    );
    expect_file_process(
        &resources,
        "src/module/init.lua",
        "local value = require(script.Parent)",
    );
}

mod sourcemap {
    use super::*;

    const CONVERT_PATH_TO_ROJO_SOURCEMAP_CONFIG: &str = r#"{
        generator: 'retain_lines',
        rules: [
            {
                rule: 'convert_require',
                current: {
                    name: 'path',
                    sources: {
                        '@pkg': './Packages'
                    }
                },
                target: {
                    name: 'roblox',
                    rojo_sourcemap: './sourcemap.json',
                }
            }
        ]
    }"#;

    fn get_resources_for_sourcemap(datamodel_case: bool) -> Resources {
        memory_resources!(
            "src/init.lua" => include_str!("../test_cases/sourcemap/src/init.lua"),
            "src/a.lua" => include_str!("../test_cases/sourcemap/src/a.lua"),
            "src/b.lua" => include_str!("../test_cases/sourcemap/src/b.lua"),
            "src/c.lua" => include_str!("../test_cases/sourcemap/src/c.lua"),

            "src/d/init.lua" => include_str!("../test_cases/sourcemap/src/d/init.lua"),
            "src/d/d1.lua" => include_str!("../test_cases/sourcemap/src/d/d1.lua"),
            "src/d/d2.lua" => include_str!("../test_cases/sourcemap/src/d/d2.lua"),

            "Packages/Package1/init.lua" => include_str!("../test_cases/sourcemap/Packages/Package1/init.lua"),
            "Packages/Package1/value.lua" => include_str!("../test_cases/sourcemap/Packages/Package1/value.lua"),

            "main.server.lua" => include_str!("../test_cases/sourcemap/main.server.lua"),

            ".darklua.json" => CONVERT_PATH_TO_ROJO_SOURCEMAP_CONFIG,
            "sourcemap.json" => if datamodel_case {
                include_str!("../test_cases/sourcemap/place-sourcemap.json")
            } else {
                include_str!("../test_cases/sourcemap/sourcemap.json")
            },
        )
    }

    #[test]
    fn invalid_sourcemap() {
        let resources = memory_resources!(
            "src/init.lua" => "return nil",
            ".darklua.json" => CONVERT_PATH_TO_ROJO_SOURCEMAP_CONFIG,
            "sourcemap.json" => "",
        );
        utils::snapshot_file_process_file_errors(&resources, "src/init.lua", "invalid_sourcemap")
    }

    #[test]
    fn convert_sibling_module_from_init_module() {
        snapshot_file_process(
            &get_resources_for_sourcemap(false),
            "src/d/init.lua",
            "convert_sibling_module_from_init_module",
        );
    }

    #[test]
    fn in_datamodel_convert_sibling_module_from_init_module() {
        snapshot_file_process(
            &get_resources_for_sourcemap(true),
            "src/d/init.lua",
            "convert_sibling_module_from_init_module",
        );
    }

    #[test]
    fn convert_module_from_child_module() {
        snapshot_file_process(
            &get_resources_for_sourcemap(false),
            "src/d/d2.lua",
            "convert_module_from_child_module",
        );
    }

    #[test]
    fn in_datamodel_convert_module_from_child_module() {
        snapshot_file_process(
            &get_resources_for_sourcemap(true),
            "src/d/d2.lua",
            "convert_module_from_child_module",
        );
    }

    #[test]
    fn convert_multiple_sibling_modules_from_root_init_module() {
        snapshot_file_process(
            &get_resources_for_sourcemap(false),
            "src/init.lua",
            "convert_multiple_sibling_modules_from_root_init_module",
        );
    }

    #[test]
    fn in_datamodel_convert_multiple_sibling_modules_from_root_init_module() {
        snapshot_file_process(
            &get_resources_for_sourcemap(true),
            "src/init.lua",
            "convert_multiple_sibling_modules_from_root_init_module",
        );
    }

    #[test]
    fn convert_sibling_module_from_sibling_module() {
        expect_file_process(
            &get_resources_for_sourcemap(false),
            "src/b.lua",
            "local a = require(script.Parent:FindFirstChild('a'))\n\nreturn a\n",
        );
    }

    #[test]
    fn in_datamodel_convert_sibling_module_from_sibling_module() {
        expect_file_process(
            &get_resources_for_sourcemap(true),
            "src/b.lua",
            "local a = require(script.Parent:FindFirstChild('a'))\n\nreturn a\n",
        );
    }

    #[test]
    fn convert_package_module_from_nested_module() {
        snapshot_file_process(
            &get_resources_for_sourcemap(false),
            "src/d/d1.lua",
            "convert_package_module_from_nested_module",
        );
    }

    #[test]
    fn in_datamodel_convert_package_module_from_nested_module() {
        snapshot_file_process(
            &get_resources_for_sourcemap(true),
            "src/d/d1.lua",
            "convert_package_module_from_nested_module",
        );
    }

    #[test]
    fn convert_nested_package_module_from_sibling_module() {
        snapshot_file_process(
            &get_resources_for_sourcemap(false),
            "Packages/Package1/init.lua",
            "convert_nested_package_module_from_sibling_module",
        );
    }

    #[test]
    fn in_datamodel_convert_nested_package_module_from_sibling_module() {
        snapshot_file_process(
            &get_resources_for_sourcemap(true),
            "Packages/Package1/init.lua",
            "convert_nested_package_module_from_sibling_module",
        );
    }

    // following tests are only on the DataModel sourcemap case
    #[test]
    fn in_datamodel_convert_module_require_across_service_instance() {
        snapshot_file_process(
            &get_resources_for_sourcemap(true),
            "main.server.lua",
            "convert_module_require_across_service_instance",
        );
    }
}
