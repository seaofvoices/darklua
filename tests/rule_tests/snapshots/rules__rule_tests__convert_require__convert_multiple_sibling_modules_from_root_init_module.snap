---
source: tests/rule_tests/convert_require.rs
expression: "process `tests/test_cases/sourcemap/src/init.lua`"
---
local b = require(script:FindFirstChild('b'))
local c = require(script:FindFirstChild('c'))
local d = require(script:FindFirstChild('d'))

local Package1 = require(script:FindFirstChild('Packages'):FindFirstChild('Package1'))

return {
    b = b,
    c = c,
    d = d,
    Package1 = Package1,
}

